README.md
# Squid Auth LDAP

### authentication via ldap for squid proxy server

I last did updated or made changes to this before 2002 (I cannot remember), I'll not look at it again, but put it here for anyone to fork and work if they feel like. Though there are better tools included in Squid these days. 

Based on squid_ldap_auth, by Glen Newton Modified to allow contextless login, to an Novell NDS LDAP Server, plus some other stuff like check for group member ship, and logging.  

Dependencies: You need to get the OpenLDAP libraries from http://www.openldap.org, or Novell's LDAP C Libraries for Linux, http://developer.novell.com

License: squid_ldap_auth is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 or (at your option) any later version.

Configuration: This is the information which has to be entered in squid.conf

```
authenticate_program /usr/local/squid/bin/squid_auth_ldap
  -S ldap.acme.com -s o=acme -g cn=users,o=acme
```
or you can use a configuration file

```
authenticate_program /usr/local/squid/bin/squid_auth_ldap \ 
  -f squid_auth_ldap.conf
```

squid_auth_ldap.conf
```
ldapserver myldap.acme.com
ldapport 389
searchbase o=acme
proxygroup cn=users,o=acme
loglevel 0
```

acl information in squid.conf
```
acl all 0.0.0.0/0.0.0.0
acl user_password proxy_auth REQUIRED
http_access allow user_password
http_access deny all
```


CHANGES
2.0.b3 2002/08/29
- code clean up (I can keep on....)
- start logging implementation (syslog)

2.0.b2 2002/04/22
- getopt
- even more clean up....

2.0.b1 2001/04/01
- added a option for a configuration file
- made install work 

2.0.b1 2001/03/26
- split code into modules

2.0.b1 2001/03/25
- added code for checking membership of a group

1..0.1 2001/03/18
- first initial release
- put program into a loop - faster authentication
- made sure that logging code is not compiled, if not needed
- added changes to README regarding configuration


main.c
/*
  main.c
   
  Author: Casper Pedersen
          casperghst42@fastmail.fm
  
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/	   

#include <stdio.h>
#include <string.h>
#include <lber.h>
#include <ldap_cdefs.h>
#include <ldap.h>
#include <stdio.h>
#include <sys/time.h>
#include <stdarg.h>
#include <unistd.h>

// --------------------------------------------------------
#include "defines.h"
// --------------------------------------------------------


// --------------------------------------------------------
int getUIDPass (char* buf, char *user, char *pass);

int main (int argc, char **argv)
{
	char dnGrp[1024];  
  char buf[256];
  char udn[256];

  LDAP *ld;

  int rc, retVal = 1;
  int chkGroup = 0;
  int opt = 0;
  struct config conf;

  setbuf(stdout,NULL);

  _StartLog( );
  logging( RUN, "%s - %s - starting", PROG, VERS );

  if( (opt = options( argc, argv, &conf )) == 0 )
  {
    exit(1);
  }

  if( opt == OK_CFG )
  {
    if( readconf( conf.confFile, &conf ) == 0 )
      exit(1);
  }

  StartLog( &conf );

	// Get FQN of group - crappy code
  if( conf.is_pxyGroup )
  {
    int ret;
    logging( RUN, "- find DN for group %s\n", conf.pxyGroup );
    ret = getGroupDN(conf.ldapServer, conf.ldapPort, conf.searchBase, conf.pxyGroup, dnGrp );
    if( ret == -1 )
      exit(1);
    else
    if( ret != 1 )
    {
      logging( WARN, "- unable to find group: %s", conf.pxyGroup );
      exit(1);
    }
    
    logging( DEBUG, "- group DN: %s", dnGrp );
    chkGroup = 1;
  }  

  logging( RUN, "%s - %s - ready", PROG, VERS );

	// Main loop
  while( fgets(buf, 256, stdin) != NULL )
  {    
		char user[256];
		char passwd[256];
    int getid;
		
		user[0] = '\0';
		passwd[0] = '\0';
				
		startMsg( );	
	
    // If we didn't get both user and passwd - try again.
    getid = getUIDPass(buf, user, passwd);

		buildMsg( " - usr %s ", user );

    if( getid == 100 )
    {
      StopLog( );
      exit(100);
    }
      
    if( getid == 0 )
    {
      logging( DEBUG, "- did not get a valide string from the browser, ie. user pass");
      continue;
    }
    
		// ldap_open is faster than ldap_init
    if ((ld = ldap_open(conf.ldapServer, conf.ldapPort)) == NULL)
    {
      logging( RUN, "- unable to connect to LDAP server: %s:%d", conf.ldapServer, conf.ldapPort);
      exit (1);
    }
    
    logging( DEBUG, "- connected to ldapServer %s:%d", conf.ldapServer, conf.ldapPort);
			
    if( findUserDN( ld, conf.searchBase, user, udn ) )
    {
		  int isInGrp = 0;
    
		  buildMsg( "-> %s ", udn );
    
		  if( chkGroup )
      {
        if( isUserInLDAPgrp( ld, dnGrp, udn ) )
          isInGrp = 1;
      }  
      else
        isInGrp = 1;
    
      if( isInGrp )
      {
        retVal = canBind (ld, udn, passwd);
        if( retVal == LDAP_SUCCESS ) 
				{
					buildMsg( "-> Authenticated!" );
				  printf(OK);
				}
        else
				{
					buildMsg( "-> Not Authenticated!");
          printf(ERR);	
				}
      }
      else
			{
				buildMsg( "-> Not in Grp %s", udn );			
        printf(ERR);
			}
    }
    else
    {
		  buildMsg( "-> Cannot find user!");
      printf(ERR);
    }      
    
		logMsg( INFO );
    ldap_unbind (ld);    
  }


  logging( RUN, "%s - %s - stopping", PROG, VERS );
  StopLog();
  return 0;
}

int getUIDPass(char *buf, char *user, char *pass)
{
  char *p, *uid, *pwd;

  if(( p = strchr(buf, '\n')) != NULL)
    *p = '\0';

  logging( DEBUG, "- user string: %s", buf);

  if ((uid = strtok(buf, " ")) == NULL)
  {
    printf(ERR);
    logging( WARN, "- ERR: user");
    return 0;
  }

#ifdef DODEBUG
  if ((strcasecmp(uid, "quit")) == 0)
  {
    logging( WARN, "- quitting...");
    return 100;
  }
#endif
    
  if ((pwd = strtok(NULL, "")) == NULL)
  {
    printf (ERR);
    logging( WARN, "- ERR: no password!");
    return 0;
  }

  strcpy (user, uid);
  strcpy (pass, pwd);
  logging( DEBUG, "- got User: %s", user );
 	logging( DEBUG, "- got Password: %s", pass );

  return 1;
}




defines.h
/*
  Author: Casper Pedersen
          casperghst42@fastmail.fm
  
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/
//---------------------------------------------------------------------------
#ifndef SQUID_AUTH_LDAP
#define SQUID_AUTH_LDAP
//---------------------------------------------------------------------------
#define PROG "squid_auth_ldap"
#define VERS "2.0.b2"
//---------------------------------------------------------------------------
#include <stdio.h>
#include <string.h>
#include <lber.h>
#include <ldap_cdefs.h>
#include <ldap.h>
#include <sys/time.h>

//---------------------------------------------------------------------------
#define OK "OK\n"
#define ERR "ERR\n"
#define USRSRCATTR "uid"
#define MEMBERATTR "uniqueMember"
//---------------------------------------------------------------------------
#define OK_NOCFG 1
#define OK_CFG 2
//---------------------------------------------------------------------------

// ldap_utils.c
int canBind( LDAP* ld, char *dn, char *passwd );

int findUserDN( LDAP* ld, char* searchBase, char *uid, char *udn );

int getGroupDN( char* ldapServer, int ldapPort, char* srcBase, 
                char* grp, char* grpDN );

int isUserInLDAPgrp( LDAP *ld, char *gdn, char *dn );

// logging.c
#define LOG_FACILITY LOG_LOCAL6 // (see "man 3 syslog")
#define DEBUG 3
#define WARN  2
#define INFO  1
#define RUN   0

// readconf.h

struct config {
  char ldapServer[200];
  int is_ldapServer;
  int  ldapPort;
  int is_ldapPort;
  char searchBase[512];
  int is_searchBase;
  char pxyGroup[512];
  int is_pxyGroup;
  char confFile[256];
  int is_confFile;
  int logLevel;
  char logFile[256];  
  int is_logFile;
  int logDebug;
};

#define LDAPSERVER "LDAPSERVER"
#define LDAPPORT   "LDAPPORT"
#define SEARCHBASE "SEARCHBASE"
#define PROXYGROUP "PROXYGROUP"
#define LOGLEVEL   "LOGLEVEL"
#define LOGFILE    "LOGFILE"

int readconf( char* conffile, struct config* );

// options.c
int options( int argc, char* argv[], struct config* );

// log
void StartLog( struct config* );
void _StartLog( );
void StopLog( );
void logging( int ll, const char* fmt, ... );
void startMsg( );
void logMsg( int ll );
void buildMsg( const char* fmt, ...);

//---------------------------------------------------------------------------
#endif
//---------------------------------------------------------------------------




logging.c
/*
  logging.c
  general function for logging
  
  Author: Casper Pedersen
          casperghst42@fastmail.fm
  
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include <stdio.h>
#include <stdarg.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <syslog.h>


#include "defines.h"

static char _msgBuffer[2048];
static int  _logLevel;

void _StartLog( )
{
  openlog(PROG, LOG_CONS | LOG_PID, LOG_FACILITY );
}

void StartLog( struct config *conf )
{
  StopLog( );
	_logLevel = conf->logLevel;
  if( conf->logDebug == 0 )
    openlog(PROG, LOG_CONS | LOG_PID, LOG_FACILITY );
  else
  if( conf->logDebug == 1 )
  {
    openlog(PROG, LOG_CONS | LOG_PERROR | LOG_PID, LOG_FACILITY );
  }  
}
  
void StopLog( )
{
  closelog( );
}

void logging( int ll, const char* fmt, ... )
{
  char buffer[1024];
  va_list ap;
  va_start( ap, fmt );
  
  vsnprintf( buffer, 1024, fmt, ap );  

  if( ll == DEBUG && _logLevel >= DEBUG )
	{
	  syslog( LOG_INFO, buffer );
#ifdef DEBUG		
		printf("DEBUG\n");
#endif		
	}
	else 
	if( ll == WARN && _logLevel >= WARN )
	{
	  syslog( LOG_INFO, buffer );
#ifdef DEBUG		
		printf("WARN\n");	
#endif		
	}
	else
	if( ll == INFO && _logLevel >= INFO )
	{
	  syslog( LOG_INFO, buffer );
#ifdef DEBUG		
		printf("INFO\n");
#endif		
	}
	else
	if( ll == RUN && _logLevel >= RUN )
	{
	  syslog( LOG_INFO, buffer );		
#ifdef DEBUG		
		printf("RUN\n");		
#endif		
	}
}

void startMsg( )
{
  _msgBuffer[0] = '\0';
}

void logMsg( int ll )
{
	logging( ll, _msgBuffer );
}

void buildMsg( const char *fmt, ... )
{
  char buffer[1024];
  va_list ap;
  va_start( ap, fmt );
  
  vsnprintf( buffer, 1024, fmt, ap );  
  strcat( _msgBuffer, buffer );  
}


readconf.c
/*
  readconf.c
  read the configuration file
  
  Author: Casper Pedersen
          casperghst42@fastmail.fm
  
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/	   
*/

#include "defines.h"

static void head( char* );
static void trail( char* );

int readconf( char* conffile, struct config* conf )
{
  FILE *fp;
  char buf[512];
  char *p;
  int  line = 0;
  int  found = 0;

  conf->is_ldapServer = 0;
  conf->is_ldapPort = 0;
  conf->is_searchBase = 0;
  conf->is_pxyGroup = 0;
  conf->is_confFile = 0;
  conf->is_logFile = 0;
  conf->logLevel = 0;
  conf->logDebug = 0;

  conf->ldapServer[0] = '\0';
	conf->ldapPort = 389;
  conf->searchBase[0] = '\0';
  conf->pxyGroup[0] = '\0';
  conf->logFile[0] = '\0';
  
  fp = fopen( conffile, "r" );
  if( fp == NULL )
  {
    logging( RUN, "- cannot open configuration file: %s", conffile );
    return 0;
  }
  
  while( (fgets(buf, 512, fp)) != NULL)
  {
    char *parm;
    char *val;
    
    line++;
    
    if( (p = strchr(buf, '\n')) != NULL )
      *p = '\0';
    
    if( buf[0] == '#' )
      continue;
      
    logging(DEBUG,"- conf. string %s", buf);
    
    parm = strtok( buf, " " );
    val = strtok( NULL, "" );
    
    
    if( (parm == NULL) || (val == NULL) )
    {
      logging(DEBUG,"- invalid entry in line: %d", line);
      fclose(fp);
      return 0;
    }

    trail( parm ); head( parm );
    trail( val ); head( val );
    logging(DEBUG,"- parm: %s - val: %s", parm, val );
  
    if( (strcasecmp(LDAPSERVER, parm)) == 0 )
    {
      strcpy( conf->ldapServer, val );
      conf->is_ldapServer = 1;
    }
    else
    if( (strcasecmp(LDAPPORT, parm)) == 0 )
    {
      conf->ldapPort = atoi( (char*) val );
      conf->is_ldapPort = 1;
    }
    else
    if( (strcasecmp(SEARCHBASE, parm)) == 0 )
    {
      strcpy( conf->searchBase, val );
      conf->is_searchBase = 1;
    }
    else
    if( (strcasecmp(PROXYGROUP, parm)) == 0 )
    {
      strcpy( conf->pxyGroup, val );
      conf->is_pxyGroup = 1;
    } 
    else
    if( (strcasecmp(LOGFILE, parm)) == 0 )
    {
      strcpy( conf->logFile, val );
    }
    else
    if( (strcasecmp(LOGLEVEL, parm)) == 0 )
    {
      conf->logLevel = atoi( val );
    }
    
  } 
  
  fclose(fp);
 
  if( !conf->is_ldapServer )
  {
    fprintf(stderr,"- missing LDAPSERVER <param> in %s\n", conffile );
    return 0;
  }
  if( !conf->is_ldapPort )
  {
    fprintf(stderr,"- missing LDAPPORT <param> in %s\n", conffile );
    return 0;
  }
  if( !conf->is_searchBase )
  {
    fprintf(stderr,"- missing SEARCHBASE <param> in %s\n", conffile );
    return 0;
  }
  
  return 1;
}

// strip spaces from beginning
void head( char *buf )
{
  int l;
  
  if( buf == NULL ) return;

  l = strlen( buf );
  
  while( (buf[0] == ' ') && (l > 0) )
  {    
    int x;
    for( x = 0; x < l; x++ )
      buf[x] = buf[x+1];
      
    buf[l-1] = '\0';  
    
    l = strlen(buf);  
  }
}

// strip spaces from end
void trail( char *buf )
{
  int l;
  int i;

  if( buf == NULL ) return;
  
  l = strlen( buf ) - 1;
  
  for( i = l; i >= 0; i-- )
  {
    if( buf[i] == ' ' )
      buf[i] = '\0';
    else
      break; 
  }
}


ldap_utils.c
/*
  ldap_utils.c 
  general functions for LDAP
	 
  Author: Casper Pedersen
          casperghst42@fastmail.fm
  
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include <stdio.h>
#include <string.h>
#include <lber.h>
#include <ldap_cdefs.h>
#include <ldap.h>
#include <sys/time.h>

#include "defines.h"

int canBind(LDAP *ld, char *dn, char *passwd)
{
  int rc; 
  logging( INFO, "- password check for %s", dn );
  rc = ldap_simple_bind_s( ld, dn, passwd );
  if( rc != LDAP_SUCCESS )
  {
    logging( DEBUG, "- password not OK!");
    return rc;
  }
	
  logging( DEBUG, "- password OK!");
  return 0;
}

int isDNGrp( char *grp )
{
  char **val;
  int x;
  
  val = ldap_explode_dn( grp, 0 );
  
  if( strchr( val[0], '=') == NULL )
  {
    logging( DEBUG, "- no full group");
    return 0;
  }
  
  if( val[0][0] != '\0' )
  {
    int i = 0;
    while( val[i] != NULL )
    {
      logging( DEBUG, "- (%d) %s", i, val[i] );
      i++;
    }
    return 1;
  }
  
  free(val);

  return 0;
}

int getGroupDN( char* ldapServer, int ldapPort, char* srcBase, 
                char* grp, char* grpDN )
{
  LDAP *ld;
  LDAPMessage *searchResult, *entry;
  struct timeval timeOut;
  char searchStr[512];
  char *dn;
  int rc;
  int found = 0;
	
  timeOut.tv_sec = 10L;
  timeOut.tv_usec = 0L;

  if( isDNGrp( grp ) )
  {
    logging( DEBUG, "- group has DN");
    strcpy( grpDN, grp );
    return 1;
  }

  logging( DEBUG, "- open connection to ldapserver: %s:%d", ldapServer, ldapPort);
	
  if( (ld = ldap_open( ldapServer, ldapPort ) ) == NULL )
  {
    logging( WARN, "- cannot login to: %s:%d", ldapServer, ldapPort);
    return -1;
  }
	
  sprintf(searchStr, "(&(objectClass=groupOfNames) (| (cn=%s)))", grp);

  logging( DEBUG, "- search for: %s", searchStr );
  	
  rc = ldap_search_st( ld,
										   srcBase,
											 LDAP_SCOPE_SUBTREE,
											 searchStr,
										   NULL, 0,
											 &timeOut,
											 &searchResult );

  logging( DEBUG, "- search done");

  if( ldap_count_entries( ld, searchResult ) == 0 )
  {
    ldap_unbind(ld);
    return 0;
  }
	
  logging( DEBUG, "- found an entry");
  
  entry = ldap_first_entry( ld, searchResult );											 
  if( entry != NULL )
  {
    dn = ldap_get_dn( ld, entry );
    strcpy( grpDN, dn );
    logging( DEBUG, "- entry found: %s", grpDN );
    free( dn );
    found = 1;
  }
					
  ldap_msgfree(searchResult);
  ldap_unbind( ld );						 
	
  return found;
}

int findUserDN( LDAP *ld, char* searchBase, char* uid, char* udn )
{
  char searchStr[512];
  LDAPMessage *searchResult, *entry;
  struct timeval timeOut;
  char *dn;
  int rc;
  int found = 0;
	
  timeOut.tv_sec = 10L;
  timeOut.tv_usec = 0L;
	
  sprintf (searchStr, "(&(objectClass=person) (| (%s=%s)))", USRSRCATTR, uid);
  logging( DEBUG, "- searchstr: %s", searchStr );
  logging( DEBUG, "- start searching for uid: %s", uid );
	
  rc = ldap_search_st (ld,
                 searchBase,
		 LDAP_SCOPE_SUBTREE,
		 searchStr, 
		 NULL, 
		 0, 
		 &timeOut, 
		 &searchResult);

  logging( DEBUG, "- search done");

  if( ldap_count_entries(ld, searchResult) == 0 )
  {
    logging( WARN, "- user \"%s\", not found!\n", uid);
    ldap_msgfree (searchResult); 
		return 0;
  }

  logging( DEBUG, "- found a entry");
    
  entry = ldap_first_entry(ld, searchResult );
	
  if( entry != NULL )
  {
    dn = ldap_get_dn(ld,entry);
		
    if( dn != NULL )
    {
      strcpy( udn, dn );
      free(dn);
			
      logging( DEBUG, "- DN found: %s", udn );			
      found = 1;
    }	
    else  
    {
      udn[0] = '\0';
      found = 0;
    }	
  }

  ldap_msgfree (searchResult);    
	
  return found;	
}

int isUserInLDAPgrp( LDAP *ld, char *gdn, char *dn )
{
  int rc;
	
  logging( DEBUG, "- is user %s in %s\n", dn, gdn );
  if( ldap_compare_s( ld, gdn, MEMBERATTR, dn ) == LDAP_COMPARE_TRUE )
  {
    logging( DEBUG, "- user \"%s\" is in Group \"%s\"", dn, gdn );
    return 1;		
  }
	
  logging( DEBUG, "- user \"%s\" is NOT in Group \"%s\"", dn, gdn );
  return 0;
}



options.c
/*
  options.c
  get command options
   
  Author: Casper Pedersen
          casperghst42@fastmail.fm
  
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/	   

#include <stdio.h>
#include <unistd.h>

#include "defines.h"

static void usage( void )
{
  fprintf(stderr,"\n");
  fprintf(stderr,"Usage: squid_auth_ldap -[f config-file] -[Sps[gld]] -h\n");
  fprintf(stderr,"       -f cofiguration file\n");
  fprintf(stderr,"       -S ldapServer\n");
  fprintf(stderr,"       -p ldapPort (default 389)\n");
  fprintf(stderr,"       -s searchBase\n");
  fprintf(stderr,"       -g proxyGroup\n");
  fprintf(stderr,"       -l logfile (not implemented - using syslog)\n");
  fprintf(stderr,"       -d loglevel <0..3> (default 0)\n");
  fprintf(stderr,"       -D debug - logs to stderr\n");
  fprintf(stderr,"       -h help\n");
  fprintf(stderr,"\n");
  fprintf(stderr,"- debug levels:\n");
  fprintf(stderr,"        0 - none\n");
  fprintf(stderr,"        1 - info\n");
  fprintf(stderr,"        2 - warnings\n");
  fprintf(stderr,"        3 - debug\n");
  fprintf(stderr,"\n");
  fprintf(stderr,"squid_auth_ldap -f /etc/squid_auth_ldap.conf\n");
  fprintf(stderr,"squid_auth_ldap -S acme_ldap -p 389 -s o=acme -g pxyUsers\n");
  fprintf(stderr,"\n");
}

int options (int argc, char* argv[], struct config* conf)
{
  int opt;
  char msg[2024];
  msg[0] = '\0';
      
  conf->is_ldapServer = 0;
  conf->is_ldapPort = 0;
  conf->is_searchBase = 0;
  conf->is_pxyGroup = 0;
  conf->is_confFile = 0;
  conf->is_logFile = 0;
  conf->logLevel = 0;
  conf->logDebug = 0;
  conf->ldapPort = 389;
	
	
  opt = getopt( argc, argv, "DSpsgfldh" );
  while( opt != -1 )
  {
    switch( opt )
    {
      case 'D': conf->logDebug = 1;
                logging(DEBUG,"- debug");
                break;
    
      case 'S': strcpy( conf->ldapServer, argv[optind] );
                conf->is_ldapServer = 1;
                logging(DEBUG,"- ldapServer: %s ", conf->ldapServer );
								break;
      
      case 'p': conf->ldapPort = atoi( (char*)argv[optind]);
                conf->is_ldapPort = 1;
                logging(DEBUG,"- ldapPort: %d ", conf->ldapPort );
								break;

      case 's': strcpy( conf->searchBase, argv[optind] );
                conf->is_searchBase = 1;
                logging(DEBUG,"- searchBase: %s ", conf->searchBase );
								break;
		
      case 'g': strcpy( conf->pxyGroup, argv[optind] );
                conf->is_pxyGroup = 1;
                logging(DEBUG,"- pxyGroup: %s ", conf->pxyGroup );
								break;

      case 'f': strcpy( conf->confFile, argv[optind] );
                conf->is_confFile = 1;
                logging(DEBUG,"- confFile: %s ", conf->confFile );
								break;

      case 'l': // strcpy( conf->logFile, argv[optind] );
                /// conf->is_logFile = 1;
                /// logging("OPT logFile: %s\n", conf->logFile );
								break;
		
      case 'd': conf->logLevel = atoi(argv[optind]);
                logging(DEBUG,"- logLevel: %d ", conf->logLevel );
								break;		
		
      case 'h': usage( );
                exit(0);
		
      default : usage( ); 
                exit(0);
    }
    opt = getopt( argc, argv, "DSpsgfldh" );
  }
  
  if( conf->is_ldapServer && conf->is_ldapPort && conf->is_searchBase )
  {
    logging(RUN,"- enough params to run!");
    return( OK_NOCFG );
  }
  else
  if( conf->is_confFile )
  {
    logging(RUN,"- enough params to run!");    
    return( OK_CFG );
  }
  
  usage( );
  
  return 0;
}
